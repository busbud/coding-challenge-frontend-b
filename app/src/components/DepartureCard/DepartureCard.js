import React from "react";

import { getTimeFromMins, getTimefromDate, formatPrice } from "./../../utils/helper"

import "./DepartureCard.scss";

const DepartureCard = ({
  departure,
  operator,
  locations

}) => {
  const {
    prices,
    departure_time,
    arrival_time,
    duration,
  } = departure;

  console.log(locations)

  return (
    <div className="departure-card">
      <div className="departure-card__header">
        <img src={operator.logo_url} alt={operator.name} />
        <p className="departure-card__price">{formatPrice(prices.total)} {prices.currency}</p>
      </div>
      <div className="departure-card__content">
        <div className="departure-card__origin">
          <div className="departure-card__time">{getTimefromDate(departure_time)}</div>
          <div className="departure-card__location">
            <span className="departure-card__location__city">{locations.origin.city.name}</span>
            <span className="departure-card__location__place">{locations.origin.name}</span>
          </div>
        </div>
        <div className="departure-card__destination">
          <div className="departure-card__time">{getTimefromDate(arrival_time)}</div>
          <div className="departure-card__location">
            <span className="departure-card__location__city">{locations.destination.city.name}</span>
            <span className="departure-card__location__place">{locations.destination.name}</span>
          </div>
        </div>
      </div>
      <div className="departure-card__footer">
        <div className="departure-card__travel-time">{getTimeFromMins(duration)}</div>
      </div>
    </div>
  )
}

export default DepartureCard;
