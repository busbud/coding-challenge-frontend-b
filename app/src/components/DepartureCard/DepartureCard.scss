@import './../../styles/shared';

.departure-card {
  background-color: $grey-light;
  border-radius: 6px;
  padding: 10px;
  

  &:not(:first-child) {
    margin: 15px 0 0;
  }

  &:not(:last-child) {
    margin-bottom: 15px;
  }
  
  &__header {
    display: flex;
    justify-content: space-between;
    align-items: center;

    img {
      max-height: 15px;
    }
  }

  &__price {
    margin: 0; 
  }

  &__content {
    margin: 15px 0;
    font-size: 0.875rem;
  }

  &__origin,
  &__destination {
    display: flex;
    margin: 10px 0;
  }

  &__time {
    color: $primary-color;
    font-weight: 600;
    margin-right: 10px;
  }

  &__location {
    display: flex;
    flex-direction: column;
    &__place {
      font-weight: 400;
      font-size: 0.75rem;
    }

    &__city {
      font-weight: 600;

    }
  }

  &__travel-time {
    color: #3b3b3b;
    font-size: 12px;
  }
  

  @include lg {

    &__header {
      img {
        max-height: 25px;
      }
    }
    
    padding: 20px 30px;

    margin-left: auto;
    margin-right: auto;
    max-width: 1100px;
  }
}