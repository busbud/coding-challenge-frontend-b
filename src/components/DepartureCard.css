.DepartureCard {
    display: block;
    position: relative;
    padding: 2.5rem;
    border-radius: 1rem;
    margin-bottom: 2rem;
    color: #313131;
    text-decoration: none;
    overflow: hidden;
    background-color:white;
    box-shadow: 0px 1px 3px rgba(206, 206, 206, 0.64);
}

.DepartureCard::before {
    position: absolute;
    left: 0;
    top: 0;
    content: "";
    width: 0.5rem;
    height: 100%;
    background-color: #9bdf56;
    background-color: #448fff;
    background-color: #facb0c;
}

.DepartureCard__time-container, .DepartureCard__footer, .DepartureCard__locations {
    display: flex;
    justify-content: space-between;
}
.DepartureCard__time-container, .DepartureCard__locations{
    margin-bottom: 1.5rem;
}

.DepartureCard__time span:first-child {
    font-size: 3.2rem;
}

.DepartureCard__time span:last-child {
    font-size: 2rem;
}

.DepartureCard__location{
    text-align: left
}
.DepartureCard__location div:first-child {
    font-size: 2rem;
}

.DepartureCard__location div:last-child {
    font-size: 1.5rem;
}
.DepartureCard__price {
    color: #127ccb;
    font-weight: bold
}
.DepartureCard__price span:first-child{
    font-size: 4.0rem;
}
.DepartureCard__price span:last-child{
    font-size: 2rem;

}
@media screen and (max-width: 480px) {

.DepartureCard__time span:first-child {
    font-size: 2.5rem;
}

.DepartureCard__time span:last-child {
    font-size: 1.5rem;
}

.DepartureCard__location div:first-child {
    font-size: 1.5rem;
}

.DepartureCard__location div:last-child {
    font-size: 1.2rem;
}
}
