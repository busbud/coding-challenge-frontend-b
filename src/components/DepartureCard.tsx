import React, { FC } from "react";
import { toCADString } from "../utils/money";
import { toLocalTime, humanize } from "../utils/time";

interface Props {
  departure: XDeparture;
  cities: City[];
  operators: Operator[];
  locations: Location[];
}

const DepartureCard: FC<Props> = ({
  departure,
  operators,
  locations,
  cities,
}) => {
  const {
    arrival_time,
    arrival_timezone,
    departure_time,
    departure_timezone,
    prices,
    duration,
    origin_location_id,
    destination_location_id,
    operator_id,
  } = departure;

  // Extract associated data from cities, locations and operators
  const price = toCADString(prices.total);
  const originLocation = locations.find((l) => l.id === origin_location_id);
  const destinationLocation = locations.find(
    (l) => l.id === destination_location_id
  );
  const operator = operators.find((o) => o.id === operator_id);
  const originCity = cities[0];
  const destinationCity = cities[1];

  // Build terminal location strings
  const originString = originCity && originLocation && (
    <span>
      {originCity.name} - {originLocation.name}
    </span>
  );
  const destinationString = destinationCity && destinationLocation && (
    <span>
      {destinationCity.name} - {destinationLocation.name}
    </span>
  );

  return (
    <div className="mx-0 p-6 w-full shadow-md max-w-screen-sm md:max-w-screen-md bg-white m-2">
      <div className="flex justify-between">
        <div className="mr-4 md:mr-20">
          {operator && (
            <img
              src={operator.logo_url}
              alt={`${operator.display_name} logo`}
              className="max-w-full"
              style={{ maxHeight: "80px" }}
            />
          )}
          <p>
            Departs{" "}
            <span className="font-semibold text-bb-blue">
              {toLocalTime(departure_time, departure_timezone)}
            </span>{" "}
            {originString}
          </p>
          <p className="text-gray-600 mb-2">
            Arrives{" "}
            <span className="font-semibold">
              {toLocalTime(arrival_time, arrival_timezone)}
            </span>{" "}
            {destinationString}
          </p>
          <p className="text-gray-600 italic">{humanize(duration, "en")}</p>
        </div>
        <div className="flex flex-col justify-between items-end">
          <p className="text-bb-blue font-bold text-xl">{price}</p>
          <button
            className="cursor-pointer bg-bb-orange text-white px-6 py-2"
            onClick={() => alert(NOPE)}
          >
            Select
          </button>
        </div>
      </div>
    </div>
  );
};

const NOPE =
  "This is Fauxsheaga, not Fyre Festival. We aren't going to charge you for a fake music extravaganza.";

export default DepartureCard;
