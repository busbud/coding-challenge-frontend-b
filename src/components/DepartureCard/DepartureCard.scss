@import "../../containers/App/App.scss";

.departure-card {
    border-radius: 0;
    border-color: lighten($color: $pumpkin, $amount: 25%);

    h5 {
        color: $pumpkin;
        font-weight: 500;
    }

    label {
        font-weight: 500;
    }
}
