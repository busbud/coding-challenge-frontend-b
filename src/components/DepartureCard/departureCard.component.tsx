import { useContext } from "react";
import { useTranslation } from "react-i18next";
import { convertPrice, parseTime } from "../../utils/utils";
import { BusbudContext } from "../../interfaces/context.interface";

const DepartureCard = ({ data }) => {
  const { currencyValue } = useContext(BusbudContext);

  const { t } = useTranslation();

  const {
    departure_time,
    arrival_time,
    prices: { total },
    trip_stops,
    origin_location_id,
    destination_location_id,
  } = data;

  const origin = trip_stops.find((stop) => {
    return stop.location_id === origin_location_id;
  });

  const destination = trip_stops.find((stop) => {
    return stop.location_id === destination_location_id;
  });

  return (
    <div className="departure-card">
      <div className="departure-card-travel-info">
        <div className="departure-card-origin">
          <span className="departure-card-origin-name">{origin.name}</span>
          <span className="departure-card-origin-time">
            {parseTime(departure_time)}
          </span>
        </div>
        <div className="departure-card-arrow">&#10230;</div>
        <div className="departure-card-destination">
          <span className="departure-card-destination-name">
            {destination.name}
          </span>
          <span className="departure-card-destination-time">
            {parseTime(arrival_time)}
          </span>
        </div>
      </div>
      <div className="departure-card-price">
        {convertPrice(total, currencyValue)}
      </div>
      <div className="departure-card-select">
        <button type="button" title={t("departures.select")}>
          {t("departures.select")}
        </button>
      </div>
    </div>
  );
};

export default DepartureCard;
