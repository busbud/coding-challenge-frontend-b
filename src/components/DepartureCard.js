import React, { memo } from 'react';
import format from 'date-fns/format';
import { useTranslation } from 'react-i18next';

import { useDataProvider } from '../contexts/SearchContext';

import './DepartureCard.css';
import { getDepartureDuration } from '../utils/time'
export const DepartureCard = memo(({ item, destinationCity, originCity, destinationLocation, originLocation }) => {
    let duration = getDepartureDuration(item);
    const { t } = useTranslation();
    return <a className="DepartureCard" href="#details" data-testid="departure-card">
        <div className="DepartureCard__time-container">
            <div className="DepartureCard__time">
                <span>{format(new Date(item.departure_time), 'HH:mm')}</span>
                <span>{format(new Date(item.departure_time), 'a')}</span>
            </div>
            <div className="DepartureCard__time hidden-sm">
                {duration.days ? duration.days + 'd' : null}
                {duration.hours ? duration.hours + 'h' : null}
                {duration.minutes ? duration.minutes + 'm' : null}
            </div>
            <div className="DepartureCard__time">
                <span>{format(new Date(item.arrival_time), 'HH:mm')}</span>
                <span>{format(new Date(item.arrival_time), 'a')}</span>
            </div>
        </div>
        <div className="DepartureCard__locations">
            <div className="DepartureCard__location">
                <div>{originCity.name}, {originCity.region.region_code}</div>
                <div>{originLocation.name}</div>
            </div>
            <div className="DepartureCard__location">
                <div>{destinationCity.name}, {destinationCity.region.region_code}</div>
                <div>{destinationLocation.name}</div>

            </div>
        </div>
        <div className="DepartureCard__footer">
            <div className="DepartureCard__details">
                {t("Details")}
            </div>
            <div className="DepartureCard__price">
                <span>{item.prices.total / 100} </span>
                <span>{item.prices.currency}</span>
            </div>
        </div>
    </a>
});

const DepartureCardConnected = ({ item }) => {
    const { cities, locations } = useDataProvider();
    const destinationLocation = locations[item.destination_location_id] || {};
    const originLocation = locations[item.origin_location_id] || {};
    const destinationCity = cities[destinationLocation.city_id];
    const originCity = cities[originLocation.city_id];

    if (!destinationCity || !originCity) {
        return <div>Loading...</div>;
    }

    return <DepartureCard
        item={item}
        destinationLocation={destinationLocation}
        originLocation={originLocation}
        destinationCity={destinationCity}
        originCity={originCity}
    />;
}

export default DepartureCardConnected;
