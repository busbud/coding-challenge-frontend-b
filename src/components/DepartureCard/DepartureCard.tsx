import React from "react";

import moment from "moment";

import "./DepartureCard.scss";
import { Card, Row, Col } from "react-bootstrap";

interface IDepartureCardProps {
  departureLocation: string;
  departureTime: string;
  arrivalLocation: string;
  arrivalTime: string;
  price: string;
}

const DepartureCard: React.FC<IDepartureCardProps> = ({
  departureLocation,
  departureTime,
  arrivalLocation,
  arrivalTime,
  price,
}) => {
  return (
    <Card
      className="m-3 departure-card"
      style={{ width: "14rem", height: "auto" }}
    >
      <Card.Body>
        <Row>
          <Col>
            <h5>{departureLocation}</h5>
            <p>{moment.utc(departureTime).format("LT")}</p>
          </Col>
          <Col>
            <h5>{arrivalLocation}</h5>
            <p>{moment.utc(arrivalTime).format("LT")}</p>
          </Col>
        </Row>
        <Row>
          <Col>
            <label>{price}</label>
          </Col>
        </Row>
      </Card.Body>
    </Card>
  );
};

export default DepartureCard;
