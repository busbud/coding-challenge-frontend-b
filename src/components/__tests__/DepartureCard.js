import React from 'react';
import DepartureCard from '../DepartureCard';
import renderer from 'react-test-renderer';

import moment from 'moment';
import momentDurationFormatSetup from 'moment-duration-format';
// adds moment duration format plugin to moment
momentDurationFormatSetup(moment);

it('renders correctly', () => {
  const tree = renderer
    .create(<DepartureCard
      departureDate="2018-07-30"
      arrivalDate="2018-07-30"
      departureLocation= "Location 1"
      arrivalLocation= "Location 2"
      price={100.32}
      duration={600}
    />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
