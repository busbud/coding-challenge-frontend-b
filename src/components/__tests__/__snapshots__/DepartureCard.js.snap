// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="box is-radiusless sc-bdVaJa dajSOh"
>
  <div
    className="level is-mobile"
  >
    <div
      className="level-item has-text-centered"
    >
      <div
        className="departure-date"
      >
        <p
          className="day"
        >
          30
        </p>
        <p
          className="month"
        >
          Jul
        </p>
        <p
          className="year"
        >
          2018
        </p>
      </div>
    </div>
    <div
      className="level-item"
    >
      <div>
        <div
          className="departure time-place"
        >
          <p
            className="time"
          >
            12:00 AM
          </p>
          <p
            className="place has-text-grey"
          >
            Location 1
          </p>
        </div>
        <div
          className="arrival time-place"
        >
          <p
            className="time"
          >
            12:00 AM
          </p>
          <p
            className="place has-text-grey"
          >
            Location 2
          </p>
        </div>
      </div>
    </div>
    <div
      className="level-item is-hidden-mobile"
    >
      <span
        className="icon is-small"
      >
        <i
          className="fa fa-clock-o has-text-grey"
        />
      </span>
      10h 00m
    </div>
    <div
      className="level-item"
    >
      <span
        className="price"
      >
        US$ 
        100.32
      </span>
    </div>
  </div>
</div>
`;
